catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/history/Historian.java

```diff
@@ -0,0 +1,512 @@
+        MetacardVersion rootVersion = new MetacardVersion(root, versionAction, getSubject.get());
+        Supplier<Stream<ContentItem>> relatedContent = () -> contentItems.stream()
+                .filter(ci -> ci.getId()
+                        .equals(id));
+
+        ContentItem rootItem = relatedContent.get()
+                .filter(ci -> !isNullOrEmpty(ci.getUri()))
+                .filter(ci -> !ci.getUri()
+                        .contains("#"))
+                .findAny()
+                .orElseThrow(() -> new RuntimeException(
+                        "Could not find root content item for: " + id));
+
+        List<ContentItem> derivedContent = relatedContent.get()
+                .filter(ci -> !ci.equals(rootItem))
+                .collect(Collectors.toList());
+
+        List<ContentItem> resultItems = new ArrayList<>();
+
+        // version root content item
+        long size = 0;
+        try {
+            size = rootItem.getSize();
+        } catch (IOException e) {
+            LOGGER.warn("Could not get size of item [{}].", rootItem.getId(), e);
+        }
+        resultItems.add(new ContentItemImpl(rootVersion.getId(),
+                Files.asByteSource(tmpContentPaths.get(rootItem.getId())
+                        .toFile()),
+                rootItem.getMimeTypeRawData(),
+                rootItem.getFilename(),
+                size,
+                rootVersion));
+
+        // version derived content items
+        for (ContentItem contentItem : derivedContent) {
+            size = 0;
+            try {
+                size = contentItem.getSize();
+            } catch (IOException e) {
+                LOGGER.warn("Could not get size of item [{}].", rootItem.getId(), e);
+            }
+            resultItems.add(new ContentItemImpl(rootVersion.getId(),
+                    contentItem.getQualifier(),
+                    new WrappedByteSource(contentItem),
+                    contentItem.getMimeTypeRawData(),
+                    contentItem.getFilename(),
+                    size,
+                    rootVersion));
+        }
+
+        return resultItems;
+    }
+
+    private Subject getSystemSubject() {
+        Subject systemSubject = Security.runAsAdmin(() -> Security.getInstance()
+                .getSystemSubject());
+        if (systemSubject == null) {
+            throw new RuntimeException("Could not get systemSubject to version metacards.");
+        }
+        return systemSubject;
+    }
+
+    private boolean doSkip(@Nullable Operation op) {
+        return !historyEnabled || ((boolean) Optional.ofNullable(op)
+                .map(Operation::getProperties)
+                .orElse(Collections.emptyMap())
+                .getOrDefault(SKIP_VERSIONING, false));
+    }
+
+    private void setSkipFlag(@Nullable Operation op) {
+        Optional.ofNullable(op)
+                .map(Operation::getProperties)
+                .map(p -> p.put(SKIP_VERSIONING, true));
```

Comment:
Since you only care about the side effect and not the previous map values, I would suggest changing this last `map` to a `forEach`.